src/Server.cpp

```diff
@@ -200,73 +313,352 @@ void Server::acceptConnection() const
+ * is properly closed.
+ */
+void Server::handleTimeout()
+{
+	for (std::map<int, Connection>::iterator iter = m_connections.begin(); iter != m_connections.end();
+		/* no increment */) {
+		time_t timeSinceLastEvent = iter->second.getTimeSinceLastEvent();
+		LOG_DEBUG << iter->second.getClient() << ": Time since last event: " << timeSinceLastEvent;
+		if (timeSinceLastEvent > m_clientTimeout) {
+			LOG_INFO << "Connection timeout: " << iter->second.getClient();
+			removeEvent(m_epfd, iter->first);
+			iter->second.closeConnection();
+			m_connections.erase(iter++);
+		} else
+			++iter;
+	}
+}
+
 bool Server::checkForCompleteRequest(int clientSock)
 {
-	const size_t headerEndPos = m_requestStrings[clientSock].find("\r\n\r\n");
+	const size_t headerEndPos = m_connectionBuffers[clientSock].find("\r\n\r\n");
 
 	return (headerEndPos != std::string::npos);
-		/*
-		headerEndPos += 4;
-		size_t bodySize = m_requestStrings[clientSock].size() - headerEndPos;
-		// FIXME: add check against default/config max body size
-		size_t contentLengthPos = m_requestStrings[clientSock].find("Content-Length");
-		size_t transferEncodingPos = m_requestStrings[clientSock].find("Transfer-Encoding");
-
-		if (contentLengthPos != std::string::npos && transferEncodingPos == std::string::npos) {
-			unsigned long contentLength
-				= std::strtoul(m_requestStrings[clientSock].c_str() + contentLengthPos + 15, NULL, 10);
-			if (bodySize >= contentLength)
+	/*
+	headerEndPos += 4;
+	size_t bodySize = m_requestStrings[clientSock].size() - headerEndPos;
+	// FIXME: add check against default/config max body size
+	size_t contentLengthPos = m_requestStrings[clientSock].find("Content-Length");
+	size_t transferEncodingPos = m_requestStrings[clientSock].find("Transfer-Encoding");
+
+	if (contentLengthPos != std::string::npos && transferEncodingPos == std::string::npos) {
+		unsigned long contentLength
+			= std::strtoul(m_requestStrings[clientSock].c_str() + contentLengthPos + 15, NULL, 10);
+		if (bodySize >= contentLength)
+			return true;
+	} else if (transferEncodingPos != std::string::npos) {
+		std::string tmp = m_requestStrings[clientSock].substr(transferEncodingPos);
+		if (tmp.find("chunked") != std::string::npos && tmp.find("0\r\n\r\n") != std::string::npos)
+			return true;
+	}
+	*/
+}
+
+/**
+ * @brief Register a virtual server.
+ *
+ * Adds a virtual server to the list of virtual servers and registers it with epoll.
+ *
+ * @param serverSock The socket of the virtual server to register.
+ *
+ * @return true if the virtual server was successfully registered, false otherwise.
+ */
+bool Server::registerVirtualServer(const Socket& serverSock)
+{
+	struct epoll_event event = {};
+	event.events = EPOLLIN;
+	event.data.fd = serverSock.fd;
+
+	if (!addEvent(m_epfd, serverSock.fd, &event)) {
+		close(serverSock.fd);
+		LOG_ERROR << "Failed to add event for " << serverSock;
+		return false;
+	}
+
+	m_virtualServers[serverSock.fd] = serverSock;
```

Comment:
Yeah. I added it for convenience but I agree that this could be dangerous. 
Would the Connection Class save the fd of the client? Or would you only have it in the map(s)?

Reply:
I would just have it in the maps as keys and the Connection class shall only retain the link to its server via the server fd.

Reply:
> shall only retain the link to its server via the server fd.

I read as meaning server socket.

I removed int fd from the socket struct. 
This had the effect that we now need to pass the fd around, since we still need this info to do work on the fd. Eg registerVirtualServer() and registerConnection() need the fd to register it with poll instance; acceptConnection() and handleConnection() need fd to accept() or recv() and send().
retrieveSocketInfo() doesn't need it since its gone from the socket struct.

Not quite sure anymore if good change. But also not opposed to it!

Reply:
Yes, I am aware that this requires additional handling, but if it prevents vulnerabilities I think it is worth it.